File: backend/app/models/scene.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Literal, Optional
from datetime import datetime

class ObjectState(BaseModel):
    id: str
    timestamp: float
    properties: Dict[str, Any]
    transition: Optional[Dict[str, Any]] = None

class ManimObject(BaseModel):
    id: str
    type: Literal["Square", "Circle", "Text", "LaTeX", "Line", "Arrow"]
    position: List[float] = Field(..., min_length=2, max_length=2)
    rotation: float = 0.0
    scale: float = 1.0
    properties: Dict[str, Any]
    states: List[ObjectState] = []
    layer: int = 0

class SceneSettings(BaseModel):
    width: int = 1920
    height: int = 1080
    backgroundColor: str = "#000000"
    frameRate: int = 60
    quality: Literal["low", "medium", "high"] = "medium"

class SceneMetadata(BaseModel):
    created: datetime
    modified: datetime
    version: str = "1.0.0"

class SceneData(BaseModel):
    id: str
    objects: List[ManimObject] = []
    settings: SceneSettings
    metadata: SceneMetadata

class AnimationRequest(BaseModel):
    scene: SceneData
    request: str = Field(..., min_length=1, max_length=1000)
    settings: Optional[Dict[str, Any]] = None

class AnimationResponse(BaseModel):
    id: str
    success: bool
    video_url: Optional[str] = None
    generated_code: Optional[str] = None
    error: Optional[str] = None
    execution_time: Optional[float] = None 

File: frontend/src/lib/types/scene.ts
export interface SceneData {
  id: string;
  objects: ManimObject[];
  settings: SceneSettings;
  metadata: {
    created: Date;
    modified: Date;
    version: string;
  };
}

export interface ManimObject {
  id: string;
  type: 'Square' | 'Circle' | 'Text' | 'LaTeX' | 'Line' | 'Arrow';
  position: [number, number];
  rotation: number;
  scale: number;
  properties: Record<string, any>;
  states: ObjectState[];
  layer: number;
}

export interface ObjectState {
  id: string;
  timestamp: number;
  properties: Partial<ManimObject>;
  transition?: TransitionConfig;
}

export interface TransitionConfig {
  duration: number;
  easing: 'linear' | 'ease-in' | 'ease-out' | 'ease-in-out';
  delay?: number;
}

export interface SceneSettings {
  width: number;
  height: number;
  backgroundColor: string;
  frameRate: number;
  quality: 'low' | 'medium' | 'high';
}

// Canvas-specific types
export interface CanvasPosition {
  x: number;
  y: number;
}

export interface DragItem {
  type: string;
  id: string;
  objectType?: ManimObject['type'];
}

// Object-specific property interfaces
export interface SquareProperties {
  sideLength: number;
  color: string;
  fillOpacity: number;
  strokeWidth: number;
  strokeColor: string;
}

export interface CircleProperties {
  radius: number;
  color: string;
  fillOpacity: number;
  strokeWidth: number;
  strokeColor: string;
}

export interface TextProperties {
  text: string;
  fontSize: number;
  fontFamily: string;
  color: string;
  weight: 'normal' | 'bold';
}

export interface LaTeXProperties {
  tex: string;
  fontSize: number;
  color: string;
}

export type ObjectProperties = 
  | SquareProperties 
  | CircleProperties 
  | TextProperties 
  | LaTeXProperties; 